src/bson_mapper/bson_streambuf.hpp

```diff
@@ -0,0 +1,76 @@
+// Copyright 2016 MongoDB Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <bsoncxx/builder/stream/document.hpp>
+#include <bsoncxx/json.hpp>
+#include <mongocxx/collection.hpp>
+
+#include <bson_mapper/config/prelude.hpp>
+
+namespace bson_mapper {
+BSON_MAPPER_INLINE_NAMESPACE_BEGIN
+
+/**
+* A streambuffer that accepts bytes of BSON data,
+* and inserts a document into the given collection when
+* all bytes are sent over.
+*/
+class BSON_MAPPER_API bson_output_streambuf : public std::streambuf {
+   public:
+    /**
+    * Constructs a new BSON Output Streambuffer
+    * that inserts documents into the given collection.
+    * @param coll A pointer to a MongoDB collection.
+    */
+    bson_output_streambuf(mongocxx::collection coll);
```

Comment:
What are your thoughts on taking a callback here rather than a collection? Something like:
```c++
using document_callback = std::function<void(bsoncxx::document::value)>;

bson_output_streambuf(document_callback cb);
```

Thoughts? (cc @adamchel)